File: android/src/main/java/io/capawesome/capacitorjs/plugins/datetimepicker/DatetimePickerPlugin.java
package io.capawesome.capacitorjs.plugins.datetimepicker;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.Date;

@CapacitorPlugin(name = "DatetimePicker")
public class DatetimePickerPlugin extends Plugin {

    public static final String TAG = "DatetimePickerPlugin";

    public static final String ERROR_MODE_INVALID = "The provided mode is invalid.";

    private DatetimePicker implementation;

    public void load() {
        DatetimePickerConfig config = getDatetimePickerConfig();
        implementation = new DatetimePicker(this, config);
    }

    @PluginMethod
    public void present(PluginCall call) {
        try {
            String format = call.getString("format", "yyyy-MM-dd'T'HH:mm:ss.sss'Z'");
            String max = call.getString("max");
            String min = call.getString("min");
            String mode = call.getString("mode", "datetime");
            String theme = call.getString("theme");
            String value = call.getString("value");
            String cancelButtonText = call.getString("cancelButtonText", "Cancel");
            String doneButtonText = call.getString("doneButtonText", "Ok");

            Date date = new Date();
            if (value != null) {
                date = DatetimePickerHelper.convertStringToDate(format, value);
            }
            Date minDate = null;
            if (min != null) {
                minDate = DatetimePickerHelper.convertStringToDate(format, min);
            }
            Date maxDate = null;
            if (max != null) {
                maxDate = DatetimePickerHelper.convertStringToDate(format, max);
            }
            PresentResultCallback resultCallback = new PresentResultCallback() {
                @Override
                public void success(Date date) {
                    String value = DatetimePickerHelper.convertDateToString(format, date);
                    JSObject result = new JSObject();
                    result.put("value", value);
                    call.resolve(result);
                }

                @Override
                public void cancel() {
                    JSObject result = new JSObject();
                    result.put("value", null);
                    call.resolve(result);
                }
            };

            if (mode.equals("datetime")) {
                implementation.presentDateTimePicker(date, minDate, maxDate, cancelButtonText, doneButtonText, theme, resultCallback);
            } else if (mode.equals("date")) {
                implementation.presentDatePicker(date, minDate, maxDate, cancelButtonText, doneButtonText, theme, resultCallback);
            } else if (mode.equals("time")) {
                implementation.presentTimePicker(date, cancelButtonText, doneButtonText, theme, resultCallback);
            } else {
                call.reject(ERROR_MODE_INVALID);
            }
        } catch (Exception ex) {
            String message = ex.getLocalizedMessage();
            Log.e(TAG, message);
            call.reject(message);
        }
    }

    private DatetimePickerConfig getDatetimePickerConfig() {
        DatetimePickerConfig config = new DatetimePickerConfig();

        String theme = getConfig().getString("theme");
        config.setTheme(theme, config.getTheme());

        return config;
    }
}

File: ios/Plugin/DatetimePickerPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(DatetimePickerPlugin)
public class DatetimePickerPlugin: CAPPlugin {
    public let errorModeInvalid = "The provided mode is invalid."

    private var implementation: DatetimePicker?

    override public func load() {
        let config = getDatetimePickerConfig()
        implementation = DatetimePicker(plugin: self, config: config)
    }

    @objc func present(_ call: CAPPluginCall) {
        let format = call.getString("format") ?? "yyyy-MM-dd'T'HH:mm:ss.sss'Z'"
        let max = call.getString("max")
        let min = call.getString("min")
        let mode = call.getString("mode", "datetime")
        let theme = call.getString("theme")
        let value = call.getString("value")
        let cancelButtonText = call.getString("cancelButtonText", "Cancel")
        let doneButtonText = call.getString("doneButtonText", "Ok")

        var date: Date = Date()
        if let value = value {
            date = DatetimePickerHelper.convertStringToDate(format, value) ?? Date()
        }
        var minDate: Date?
        if let min = min {
            minDate = DatetimePickerHelper.convertStringToDate(format, min)
        }
        var maxDate: Date?
        if let max = max {
            maxDate = DatetimePickerHelper.convertStringToDate(format, max)
        }

        let completion: (Date?) -> Void = { date in
            var value: String?
            if let date = date {
                value = DatetimePickerHelper.convertDateToString(format, date)
            }
            var result = JSObject()
            result["value"] = value
            call.resolve(result)
        }

        if mode == "datetime" {
            implementation?.presentDatetimePicker(date: date, minDate: minDate, maxDate: maxDate,
                                                  cancelButtonText: cancelButtonText, doneButtonText: doneButtonText, theme: theme, completion: completion)
        } else if mode == "date" {
            implementation?.presentDatePicker(date: date, minDate: minDate, maxDate: maxDate,
                                              cancelButtonText: cancelButtonText, doneButtonText: doneButtonText, theme: theme, completion: completion)
        } else if mode == "time" {
            implementation?.presentTimePicker(date: date, cancelButtonText: cancelButtonText, doneButtonText: doneButtonText, theme: theme, completion: completion)
        } else {
            call.reject(errorModeInvalid)
        }
    }

    private func getDatetimePickerConfig() -> DatetimePickerConfig {
        var config = DatetimePickerConfig()

        if let theme = getConfigValue("theme") as? String {
            if let convertedTheme = DatetimePickerHelper.convertStringToTheme(theme) {
                config.theme = convertedTheme
            }
        }

        return config
    }
}
